### PR TITLE
Feature/platform 172 document tagging

### PR DESCRIPTION
**Pull request by solomontang**
- Document tagging for access request attachments. Users can select from a pre-determined list or type to create their own custom tags.
image
image
image

### CODE REVIEW

File: src/containers/access/src/AttachmentItem.js [hunk 1]
```diff
@@ -54,18 +65,34 @@ const ImagePreview = styled.img`
...
 const AttachmentItem = ({
   divider,
   file,
   onDelete
 } :Props) => {
-
+  const dispatch = useDispatch();
+  const [editing, setEditing] = useState(false);
+  const dateTime = getPropertyValue(file, [DATE_TIME, 0]);
+  const dateStr = DateTime.fromISO(dateTime).toLocaleString(DateTime.DATE_SHORT);
```
Comment (UnsungHero97) [inline src/containers/access/src/AttachmentItem.js:90]: No change required, just letting you know this exists:

```
import { DateTimeUtils } from 'lattice-utils';

const dateStr = DateTimeUtils.formatAsDate(dateTime);
```

File: src/containers/access/src/AttachmentItem.js [hunk 2]
```diff
@@ -79,14 +106,32 @@ const AttachmentItem = ({
...
     <ListItem divider={divider}>
       <ListItemAvatar>
         {imagePreview}
       </ListItemAvatar>
-      <ListItemText primary={name} secondary={dateStr} />
-
+      <ItemTextWrapper paddingRight="96px">
+        <ListItemText
+            primary={name}
+            secondary={!editing ? <SecondaryText date={dateStr} tag={tag} /> : null} />
```
Comment (UnsungHero97) [inline src/containers/access/src/AttachmentItem.js:126]: I prefer `editing` over `!editing`, easy to miss that `!`

```suggestion
            secondary={editing ? null : <SecondaryText date={dateStr} tag={tag} />} />
```

File: src/containers/access/src/SelectTags.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+// @flow
+import { Creatable } from 'lattice-ui-kit';
+
+import { TAG_OPTIONS } from './constants';
+
+type Props = {
+  index :?number | string;
+  onTagChange :Function;
```
Comment (UnsungHero97) [inline src/containers/access/src/SelectTags.js:8]:
```suggestion
  onTagChange :(index :?number | string, value :string) => void;
```

File: src/containers/access/src/constants.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+// @flow
+const CONSENT_FORM :'Consent Form' = 'Consent Form';
+const CPHS_REVIEW :'CPHS Review' = 'CPHS Review';
+const DATA_RELEASE_LETTER :'Data Release Letter' = 'Data Release Letter';
+const DATA_SECURITY_LETTER :'Data Security Letter' = 'Data Security Letter';
+const DATA_USE_AGREEMENT :'Data Use Agreement' = 'Data Use Agreement';
+const DEPARTMENT_SUPPORT_LETTER :'Department Support Letter' = 'Department Support Letter';
+const IRB_APPROVAL :'IRB Approval' = 'IRB Approval';
+const PARTICIPANTS_BOR :"Participant's Bill of Rights" = "Participant's Bill of Rights";
```
Comment (UnsungHero97) [inline src/containers/access/src/constants.js:9]: No change required, just a thought. These will eventually need to be tied to the specific type of access request, and will probably need to be managed.

File: src/containers/access/src/sagas/getAttachments.js [hunk 5]
```diff
@@ -65,10 +67,12 @@ function* getAttachmentsWorker(action :SequenceAction) :Generator<any, any, any>
...
       .map((file) => file.get('neighborDetails'));
 
-    response.data = fileData;
+    const filesById = Map(fileData.map((v) => [getEntityKeyId(v), v]));
+
+    response.data = filesById;
 
     yield put(getAttachments.success(action.id, {
-      data: fileData
+      data: filesById
```
Comment (UnsungHero97) [inline src/containers/access/src/sagas/getAttachments.js:75]: No change required. You could have just done `.success(action.id, filesById)`.

File: src/containers/access/src/sagas/updateAttachmentTag.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        entitySetId,
+        entities,
+        updateType: UpdateTypes.PartialReplace,
+      }),
+    );
+
+    if (updateAccessResponse.error) throw updateAccessResponse.error;
+    response = { data: updateAccessResponse.data };
+
+    yield put(updateAttachmentTag.success(action.id));
```
Comment (UnsungHero97) [inline src/containers/access/src/sagas/updateAttachmentTag.js:74]: Just checking... you don't need to pass data to `.success()`?

File: src/containers/access/src/reducers/updateAttachmentTagReducer.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+const { REQUEST_STATE } = ReduxConstants;
+
+export default function updateAttachmentTagReducer(state :Map<*, *>, action :SequenceAction) {
+
+  return updateAttachmentTag.reducer(state, action, {
+    REQUEST: () => {
+      const { path, tag } = action.value;
+      return state
+        .setIn(path, tag)
+        .setIn([UPDATE_ATTACHMENT_TAG, REQUEST_STATE], RequestStates.PENDING);
```
Comment (UnsungHero97) [inline src/containers/access/src/reducers/updateAttachmentTagReducer.js:22]: It's fine to update the store with the new value before the request finishes, but you're not handling the failure case.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: src/containers/access/src/reducers/updateAttachmentTagReducer.js (hunk 7 above)
Comment (solomontang) [reply src/containers/access/src/reducers/updateAttachmentTagReducer.js:22]: yeah i feel like the failure case should yield a snackbar but not sure yet. reverting back to previous tag doesn't seem right. just leaving for now.

Comment (UnsungHero97) [reply src/containers/access/src/reducers/updateAttachmentTagReducer.js:22]: I like the idea of toasts, let's work with Shireen to standardize usage.

**Review by UnsungHero97 — APPROVED** [review]

**Review by giuliacampana — APPROVED** [review]

**Pull request merged**